Global logging format changes

peytondmurray opened the pull request:
## Why are these changes needed?

This PR changes how the logging configuration for Ray is set, and changes the format of log messages.

After a discussion with @rkooo567 we've decided to split the logging changes into multiple PRs. This is the first in a series which makes changes to library-level logging code for Ray.

### Changes

* Attempts to consolidate logging configuration by introducing reasonable defaults in `ray/log.py`.
* This new logging configuration is done once in `ray/__init__.py` at the top of the module. Subsequent calls to the configuration are ignored.
* A logger for `ray.rllib` is configured at the WARN level, to address https://github.com/ray-project/ray/pull/31858. With this change,
https://github.com/ray-project/ray/pull/31858 can be reverted, again simplifying and consolidating logging configuration.
* Modified `test_output.py::test_logger_config` to test only the logger config, not launch a ray cluster. The test was failing intermittently, I think due to a race condition between the launch of the cluster and the reading of the subprocess's stdout, and anyway it wasn't necessary to call `ray.init` here to check that logging was configured correctly.
* Modified `python/ray/tune/tests/test_commands.py::test_ls_with_cfg` to test the underlying data, not what gets printed to stdout (which has changed with the new logging system).
* Modified a logging message in `ray.tune.automl.search_policy.AutoMLSearcher.on_trial_complete`, which in certain cases emits a logging message which tries to format a `NoneType` into a `%f` during log message formatting. This was a previously-undetected bug which showed up because the default log level is now `INFO`. This fixes a test that was failing in `test_automl_searcher.py::AutoMLSearcherTest`.

## Related issue number

Partially addresses #30005.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

peytondmurray commented:
## Failing tests

Currently there is an issue that causes failures because `pandas` is not installed on the CI runners. This affects every minimal install test suite. 

Default install is also failing, but there's a fix for this available here: #32559.

rkooo567 commented:
Please add me to the assignee section!

peytondmurray commented:
:point_up: Sorry about that! I saw your name in the assignee section, I should have added you to the reviewers as well!

peytondmurray commented:
Rebased onto `master` to pull in the new test fixes for rllib.

rkooo567 commented on doc/source/ray-observability/ray-logging.rst:
```diff
@@ -2,7 +2,30 @@
+
+Internal Ray Logging Configuration
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+When ``import ray`` is executed, Ray's logging system is initialized, generating a sensible configuration given in ``python/ray/log.py``. The default logging level is ``logging.INFO``.
```
```suggestion
When ``import ray`` is executed, Ray's logger is initialized, generating a sensible configuration given in ``python/ray/log.py``. The default logging level is ``logging.INFO``.
```

rkooo567 commented on python/ray/log.py:
```diff
@@ -0,0 +1,195 @@
+import functools
```
Don't create log.py and can you just merge it into ray_logging.py? It is confusing we have 2 files like this

rkooo567 commented on python/ray/_private/test_utils.py:
```diff
@@ -418,16 +437,19 @@ def run_string_as_driver_nonblocking(driver_script, env: Dict = None):
-    return proc
+    # Rich log handler is intended for human-readability. There are a lot of tests
+    # which inspect subprocess output for validity, so force plain logging here.
+    with temporary_environment_append({"FORCE_PLAIN_LOGGING": "1"}):
```
Instead of this, can you just check the env var from the logger or sth? 

We export `BUILDKITE="true"` to the CI tests

peytondmurray replied:
Done.

rkooo567 commented on python/ray/log.py:
```diff
@@ -0,0 +1,195 @@
+        # Lazily load ray when this is called to avoid circular import
+        import ray
+
+        if hasattr(ray, "_private"):
```
whens' the case it is false?

rkooo567 replied:
Can you add a comment to the code?

peytondmurray replied:
This has been removed as per the request below.

rkooo567 replied:
seems like it is not removed?

rkooo567 commented on python/ray/log.py:
```diff
@@ -0,0 +1,195 @@
+            logging.StreamHandler.emit(self, record)
+
+
+def call_once(func: Callable[[None], None]):
```
hmm why not just using global variable ?

rkooo567 replied:
logger_initialized = False

def f():
    global logger_initialized
    if logger_initialized:
        return

peytondmurray replied:
Well, I was thinking that a decorator would be more reusable, and because of concerns about polluting the namespace with global stuff. It also encapsulated the reconfiguration debug message nicely. I'm happy to change this though if there's already an established precedent.

rkooo567 commented on python/ray/log.py:
```diff
@@ -0,0 +1,195 @@
+        super().__init__(stream=sys.stdout)
+
+        # Needed because the existing logging system causes workers to write to
+        # a file that is read by the driver. This will need to be removed when
```
> This will need to be removed when
        # a proper logging system is set up in a way that lets workers pass
        # LogRecords to the driver.

Can we remove this comment? Still not sure if we will ever do this

peytondmurray replied:
:+1:

rkooo567 commented on python/ray/log.py:
```diff
@@ -0,0 +1,195 @@
+        },
+    }
+
+    dictConfig(
```
how does this work? what's going to happen if user calls the same thing with version 1?

peytondmurray replied:
See https://docs.python.org/3/library/logging.config.html#dictionary-schema-details - python's logging system requires us to pass a `version` key, and its value must be `1` :shrug:.

rkooo567 reviewed (commented):
Can you add unit tests? We should make sure

1. The context is properly printed within each library
2. make sure the logging_level and logging_format works after this PR

I will also play with this log once I come back from OOO (next Mon)

rkooo567 commented on doc/source/ray-observability/ray-logging.rst:
```diff
@@ -2,7 +2,30 @@
+.. code-block:: python
+
+   import logging
+   logger = logging.getLogger("ray.rllib")
```
Can you add examples for all libraries (data, tune, rllib, air, train, workflow)?

peytondmurray replied:
Yep, I added a section showing how to grab handles for each of the ray loggers, and how to set the logging level and add a handler for two different ray subpackages.

rkooo567 commented on python/ray/log.py:
```diff
@@ -0,0 +1,158 @@
+                and self.global_worker_mode == ray._private.worker.WORKER_MODE
+            ):
+                record.node_ip_address = (
+                    ray.runtime_context.get_runtime_context().worker.node_ip_address
```
remove it? Seems like it is unused?

peytondmurray replied:
Yes, this is left over from splitting this PR into two. Thanks for catching this!

rkooo567 replied:
Hmm this still seems to be not removed yet? Is it expected?

rkooo567 commented on python/ray/log.py:
```diff
@@ -0,0 +1,158 @@
+        return True
+
+
+class AnsiStripFormatter(logging.Formatter):
```
Hmm I don't see this is used anywhere?

peytondmurray replied:
Again this was left over from splitting the PR. Removed.

rkooo567 replied:
not removed?

rkooo567 commented on python/ray/log.py:
```diff
@@ -0,0 +1,158 @@
+    """A plain log handler."""
+
+    def __init__(self):
+        super().__init__(stream=sys.stdout)
```
Maybe we should just use stderr which is the default of regular stream handler?

peytondmurray replied:
Currently we use the logger to provide useful runtime information to the users rather than error reporting, so I'm inclined to keep it `stdout`. What do you think? I'm happy to switch over if you'd prefer.

peytondmurray replied:
Switched to `stderr`.

rkooo567 commented on python/ray/log.py:
```diff
@@ -0,0 +1,158 @@
+    def emit(self, record: logging.LogRecord):
+        """Emit the log message.
+
+        If this is a worker, serialize the record and send it to the driver via TCP.
```
The docstring seems wrong

peytondmurray replied:
Removed.

rkooo567 replied:
also seems to be still here?

rkooo567 commented on python/ray/log.py:
```diff
@@ -0,0 +1,158 @@
+
+def generate_logging_config():
+    """Generate the default Ray logging configuration."""
+    global logger_initialized
```
Maybe we need to lock this?

peytondmurray replied:
Possibly? It's only called once, currently, at `ray` module initialization time. The previous `call_once` decorator effectively implemented a lock, but I can add one here too.

rkooo567 replied:
yep can you add a lock just in case? should be a very simple to add one

rkooo567 commented on python/ray/log.py:
```diff
@@ -0,0 +1,158 @@
+    }
+    filters = {"context_filter": {"()": ContextFilter}}
+    handlers = {
+        "null": {"class": "logging.NullHandler"},
```
do we need this line?

peytondmurray replied:
No, I think this is okay to remove.

rkooo567 replied:
still not removed

rkooo567 commented on python/ray/log.py:
```diff
@@ -0,0 +1,158 @@
+import logging
```
Hmm can you at least move this to _private repo? Or is it also difficult?

peytondmurray replied:
Nope, there's no problem with doing this - as long as there are no calls to `import ray` inside the subpackage at initialization time, which there aren't. I've moved the file there now.

rkooo567 reviewed (commented):
A couple last comments. I will also test this on my local machine tmrw

peytondmurray commented:
After a discussion with @rkooo567, I've modified `ray.init()` to strip the handlers off of the `ray` logger if `configure_logging == False`, which should match the current implementation.

rkooo567 commented on python/ray/log.py:
```diff
@@ -0,0 +1,158 @@
+import logging
+import re
```
Move this file to _private?

rkooo567 reviewed (commented):
LInt failure. Also, I see some comments were not addressed + agent changes are not here? Is it expected? Maybe something has been screwed by rebase?

peytondmurray commented:
:point_up: Yep, this got screwed up by a rebase - there was another merge conflict, and I must have messed this up. Will fix this now.

rkooo567 commented on python/ray/tests/test_runtime_env_working_dir_2.py:
```diff
@@ -147,7 +150,6 @@ def test_large_dir_upload_message(start_cluster, option):
         output = run_string_as_driver(driver_script)
         assert "Pushing file package" in output
         assert "Successfully pushed file package" in output
-        assert "warning" not in output.lower()
```
hmm maybe we shouldn't delete this line?

rkooo567 reviewed (approved):
Before merging, we need approval on code change from rllib, tune, and serve. 

cc @sven1977 for the rllib change (especially the logging level related since it was regressed last time)

cc @krfricke for the tune test change

cc @sihanwang41 for the serve test changes.

sven1977 commented on doc/source/ray-observability/ray-logging.rst:
```diff
@@ -2,7 +2,43 @@
+
+Internal Ray Logging Configuration
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+When ``import ray`` is executed, Ray's logger is initialized, generating a sensible configuration given in ``python/ray/_private/log.py``. The default logging level is ``logging.INFO``.
```
Shouldn't the default logging level be "WARNING"? "INFO" is normally quite chatty (at least for RLlib).

peytondmurray replied:
[The default logging level is `INFO`](https://github.com/peytondmurray/ray/blob/76c47941459f68e074e0a01e0d1c34e0d8c5d8ec/python/ray/_private/log.py#L111) - except for RLlib, [which is `WARNING`](https://github.com/peytondmurray/ray/blob/76c47941459f68e074e0a01e0d1c34e0d8c5d8ec/python/ray/_private/log.py#L117) because of the issue raised in [this PR](https://github.com/ray-project/ray/pull/31858). Maybe I should add a note about this to the documentation?

We currently have a test to ensure that the logging level of ray and each subpackage is not inadvertently modified, so that this doesn't happen in the future either.

ollie-iterators commented:
Even if it didn't show up in this PR, it would be nice to have the lint tests pass for the PR.

rkooo567 commented:
cc @sven1977 @krfricke @sihanwang41 can you guys review the PR? We hope to merge the PR asap.

krfricke reviewed (approved):
Tune changes LGTM

sven1977 reviewed (approved):
Looks good to me now.
Thanks for all the hard work @peytondmurray ! :)

rkooo567 commented:
@peytondmurray we should respect the logger configured from the library if it is not done that way.

peytondmurray commented:
Ahh, this was added 2 hours ago, so I didn't see this until now.

> Does it mean lib level can't configure their logging config?

No, all it means is that you can't call `ray._private.log.generate_logging_config` more than once. Any change you make on your own (i.e. anything done outside that function) is respected.

That said, my initial feeling is to move that serve logging configuration stuff to `python/ray/_private/log.py`. Part of the goal of the original logging RFC was to unify logging configuration because it was spread across disparate parts of the code base, making it very difficult to track, modify, and debug. I'd make the case to consolidate logging configuration here, and advise against letting logging configuration leak to different parts of the code base where it gets lost. [We're already doing this with `ray.rllib`](https://github.com/peytondmurray/ray/blob/76c47941459f68e074e0a01e0d1c34e0d8c5d8ec/python/ray/_private/log.py#L116), so we're setting a precedent here. On top of that, if a special handler with emit-time context information is needed, we can do that with the same mechanisms we are already using when we handle plain logging configuration from workers.

rkooo567 commented:
cc @sihanwang41 since your PR has been merged, let's see if this PR will break anything? @peytondmurray can you merge the latest master last time? I think once you can verify the serve logger works, we can get approval and merge it. @sihanwang41 what test should he run to verify this?

ollie-iterators commented:
If things fail by rebasing to master, it will definitely not be because of this PR. There are major issues with ray happening right now.

peytondmurray commented:
Rebased and pushed, there were no conflicts. The last version I pushed was passing all tests*, so if this is failing I'm hoping it's not my fault.

*All non-flaky or broken tests, anyhow

rkooo567 commented:
since master is crazy now, let's wait a bit. Also I'd like to merge this PR after 2.4 branch cut...

angelinalg commented on doc/source/ray-observability/ray-logging.rst:
```diff
@@ -2,7 +2,43 @@
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+When ``import ray`` is executed, Ray's logger is initialized, generating a sensible configuration given in ``python/ray/_private/log.py``. The default logging level is ``logging.INFO``.
+
+All ray loggers are automatically configured in ``ray._private.ray_logging``. To change the ray library logging config:
```
```suggestion
All ray loggers are automatically configured in ``ray._private.ray_logging``. To change the Ray library logging configuration:
```

angelinalg commented on doc/source/ray-observability/ray-logging.rst:
```diff
@@ -2,7 +2,43 @@
+   logger = logging.getLogger("ray")
+   logger # Modify the ray logging config
+
+Similarly, to modify the logging configuration for any ray subcomponent, specify the appropriate logger name:
```
```suggestion
Similarly, to modify the logging configuration for any Ray subcomponent, specify the appropriate logger name:
```

angelinalg commented on doc/source/ray-observability/ray-logging.rst:
```diff
@@ -97,10 +133,9 @@ Limitations:
 Tip: To avoid `print` statements from the driver conflicting with tqdm output, use `ray.experimental.tqdm_ray.safe_print` instead.
 
-How to set up loggers
+Customizing worker loggers
```
```suggestion
Customizing Worker Loggers
```

angelinalg commented on doc/source/ray-observability/ray-logging.rst:
```diff
@@ -164,11 +201,11 @@ Logging directory structure
 .. note::
 
-    The default temp directory is ``/tmp/ray`` (for Linux and Mac OS). If you'd like to change the temp directory, you can specify it when ``ray start`` or ``ray.init()`` is called. 
+    The default temp directory is ``/tmp/ray`` (for Linux and Mac OS). If you'd like to change the temp directory, you can specify it when ``ray start`` or ``ray.init()`` is called.
```
```suggestion
    The default temp directory is ``/tmp/ray`` (for Linux and MacOS). To change the temp directory, specify it when you call ``ray start`` or ``ray.init()``.
```

angelinalg commented on doc/source/ray-observability/ray-logging.rst:
```diff
@@ -193,7 +230,7 @@ Here's a Ray log directory structure. Note that ``.out`` is logs from stdout/std
 - ``runtime_env_setup-[job_id].log``: Logs from installing :ref:`runtime environments <runtime-environments>` for a task, actor or job.  This file will only be present if a runtime environment is installed.
 - ``runtime_env_setup-ray_client_server_[port].log``: Logs from installing :ref:`runtime environments <runtime-environments>` for a job when connecting via :ref:`Ray Client <ray-client-ref>`.
-- ``worker-[worker_id]-[job_id]-[pid].[out|err]``: Python/Java part of Ray drivers and workers. All of stdout and stderr from tasks/actors are streamed here. Note that job_id is an id of the driver.- 
+- ``worker-[worker_id]-[job_id]-[pid].[out|err]``: Python/Java part of Ray drivers and workers. All of stdout and stderr from tasks/actors are streamed here. Note that job_id is an id of the driver.-
```
```suggestion
- ``worker-[worker_id]-[job_id]-[pid].[out|err]``: Python/Java part of Ray drivers and workers. All stdout and stderr from tasks or actors are streamed here. Note that job_id is an id of the driver.
```

peytondmurray commented:
I rebased onto `master` again, resolving merge conflicts. Thanks again @angelinalg. @xwjiang2010 I also incorporated your latest changes to the logging system, switching to use `logging._StderrHandler` from `logging.StreamHandler`.
